### PR TITLE
Validate source of an index in LuceneChangesSnapshot

### PR DESCRIPTION
**Pull request by dnhatn**
Today it's possible to encounter an Index operation in Lucene whose _source is disabled, and _recovery_source was pruned by the MergePolicy.  If it's the case, we create a Translog#Index without source and let the caller validate it later. However, this approach is challenging for the caller.

1. Deletes and No-Ops don't allow invoking "source()" method. The caller has to make sure to call "source()" only on index operations. The current implementation in CCR does not follow this and fail to replica deletes or no-ops.

2. It's easier to reason if a Translog#Index always has the source.

### CODE REVIEW

**Review by s1monw — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java [hunk 1]
```diff
@@ -258,10 +258,19 @@ private TopDocs searchOperations(ScoreDoc after) throws IOException {
                 assert assertDocSoftDeleted(leaf.reader(), segmentDocID) : "Delete op but soft_deletes field is not set [" + op + "]";
             } else {
                 final BytesReference source = fields.source();
+                if (source == null) {
+                    if (requiredFullRange) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java]: why is this tied to `requiredFullRange` so deep?  

PS - I thought we were heading towards always requiring full ranges? this make this problem moot.

**Review by bleskes — APPROVED** [review]
LGTM, though I asked a question..

**Comment by bleskes** [comment]
Recovery also requires a full range, we don’t enforce it on this level
because of simplicity.  Check that we wait for the local checkpoint to go
above the max seq no we sample and that we validate it when shipping.

**Comment by dnhatn** [comment]
@bleskes I see what you meant. Once this TODO https://github.com/elastic/elasticsearch/blob/ccr/server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java#L167 gets in, I will remove this leniency.

**Comment by dnhatn** [comment]
@elasticmachine test this please

**Comment by dnhatn** [comment]
@elasticmachine retest this please

**Comment by dnhatn** [comment]
Thanks @s1monw and @bleskes for reviewing

**Pull request merged**